### ISSUE TITLE
Upgrade django-textclassifier

### ISSUE THREAD

**Issue by agjohnson** [rtfd/readthedocs.org#3682]
We are currently using an older version of this package. This old version uses a central classifier file that is shared across all fields that use this validation. The file is also difficult to update, so we haven't been updating this.

The new version makes several changes:

* Moves classifier seed to database backed
* Each field has it's own classifier seeding, not shared anymore
* Replaces textblob with sklearn classifier for large speed improvement (introduces a dependency on scikit for RTD)

I'd like to get some thoughts on the code changes there so I could make a proper 2.0 release. This should theoretically drop in without many changes, but would be a good spot to test.

Code here: https://github.com/agjohnson/django-textclassifier

### Requirements

I think applying this to Project and Tag makes sense, these are both commonly targeted in spam projects. Implementation should be easy, we'll want to benchmark this afterwards though.

- [ ] Determine what fields get a classifier -- Project.name, Project.repo?, Tag.name?
- [ ] Add validator to field, avoid making a migration and changing field type
- [ ] Alter forms for Project & Tag if necessary, it might not be
- [ ] Benchmark test this is a reasonable amount of trained data (this might be fairly high, 500 projects or so, depending on what classification with multiple fields looks like
- [ ] Benchmark test this with an unreasonable amount of trained data

**Comment by davidfischer** [comment]
This looks pretty good to me. I did have a few notes:

* Unless I'm mistaken, it looks like the training data is loaded out of the DB each time we classify anything (https://github.com/agjohnson/django-textclassifier/blob/37bbd39/textclassifier/classifier.py#L61). Is there a lot of it? Will there be performance issues with that?
* Is the idea behind keeping the training data in the database to not get into a flagging arms race where you are adapting to what they do and they adapt to you? This would make it tougher to adapt for them since they don't know the training data.
* If we remove the description (#3689), what will we classify on?

**Comment by humitos** [comment]
I've reading the source code and I liked it. I proposed three very simple PR.

One thing I would like to take a look is the migration of the data we already collected in the backend we are using currently. Otherwise, we will start from scratch.

In case we can't migrate the data we have now, I'd suggest to do a simple script that reads our current database and does some dumb test (`project.name.len < 30`) to populate the field data.

**Comment by humitos** [comment]
> If we remove the description (#3689), what will we classify on?

I think this should be used in the `project.name` field instead. In fact, I don't know why we are using it in the `description` field if the spammer can create a project without filling the description at all :)

**Comment by humitos** [comment]
> Unless I'm mistaken, it looks like the training data is loaded out of the DB each time we classify anything (https://github.com/agjohnson/django-textclassifier/blob/37bbd39/textclassifier/classifier.py#L61). Is there a lot of it? Will there be performance issues with that?

I'm not sure that we have to worry about this... yet. I did a simple test with 10000 elements on a list with each item a random number of :

```
In [26]: l = []
    ...: for x in range(10000):
    ...:     l.append((random.randint(1, 1000) * 'hello ', x % 2))
    ...:         

In [27]: s = json.dumps(l)

In [28]: %timeit json.loads(s)
30.6 ms ± 375 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [29]:
```

If the `json.loads` is what you were thinking. Regarding the `fit_transform` and the `fit` method; they maybe could be refactored to not loop twice over the same list. Also, both list could be a `set()` instead of a simple list --but I'm not sure that it's a type accepted.

**Comment by agjohnson** [comment]
These are some good points. Some thoughts:

* The training data will accumulate as we classify, but the validator doesn't make any attempt to limit the number of items trained. This could be added as an addition to the classifier potentially. The problem with this isn't necessarily the growth of the field, as these will potentially only be called on Project.save and Tag.save, which should be fairly infrequent. One of the larger problems with unbounded growth on the trained data is that this will become computationally expensive. With the current version, we see considerable slow down around 50 project's worth of data. This will warrant some benchmarking.
* The point of database backed is to make updating the trained set easier, but also to allow for multiple sets of data. Right now, the version we use only has one set of data, and it's outdated.
* We can still classify on Project.name and Tag.name, or anything else that is user input and is tokenizeable.

I updated this with some actual requirements, I think load testing the maximum size of our trained set will be a good first step. I also opened some issues on the textclassifier to address some of these other issues potentially.
